LibWeb: Implement OffscreenCanvasRenderingContext2D and Integrate OffscreenCanvas

me-it-is opened the pull request.

AtkinsSJ reviewed (changes requested):
I'm not especially familiar with canvas, but I've given this a look through.

me-it-is commented on Libraries/LibWeb/HTML/OffscreenCanvasRenderingContext2D.cpp:
```diff
@@ -85,81 +122,417 @@ OffscreenCanvas const& OffscreenCanvasRenderingContext2D::canvas_element() const
+    path.line_to(bottom_right);
+    path.line_to(bottom_left);
+    path.line_to(top_left);
+    return path;
```
I just pushed a refactor, does the final state look good, before i go back to fixup the history?

AtkinsSJ reviewed (changes requested):
I do see a couple of issues with this:
1. You've moved a lot of code into header files that doesn't need to be there. https://github.com/LadybirdBrowser/ladybird/commit/64f438857b28aecc6d0f8b88d53a5f1dda2f022c recently did the opposite to try and reduce build times.
2. I think moving functionality into `CanvasDrawPath`, `CanvasRect`, etc instead keeps things closer to the spec.

me-it-is committed:
LibWeb: Move CanvasPathDrawingStyle to a cpp file

me-it-is committed:
LibWeb: Throw error when geting the context of a placeholder canvas

me-it-is committed:
LibWeb: Refactor Canvas and OffscreenCanvas to use common rendering base

Creates AbstractRenderingContext2d to serve as a common base for
both CanvasRenderingContext2d and OffscreenCanvasRenderingContext2d.

gmta commented on Libraries/LibWeb/HTML/HTMLCanvasElement.cpp:
```diff
@@ -367,6 +388,35 @@ WebIDL::ExceptionOr<void> HTMLCanvasElement::to_blob(GC::Ref<WebIDL::CallbackTyp
+    m_context = offscreen_canvas;
+
+    // 5. Set the offscreenCanvas's inherited language to the language of this canvas element.
+    // offscreen_canvas.
```
Sam is referring to this line:

```cpp
// offscreen_canvas.
```

This is not in the spec and it doesn't add anything, so it should be removed.

me-it-is replied:
I see. its fixed now.

kalenikaliaksandr reviewed (changes requested):
I only glanced through the changes. The refactoring to share a common base between the two contexts looks like a reasonable direction, but if you are interested in OffscreenCanvas support, then we should start by figuring out how to share the underlying canvas surface between the WebContent and WebWorker processes, since, in my understanding, that's mainly how it's used on real websites.

me-it-is committed:
LibWeb: Move CanvasTransform to a cpp file

me-it-is committed:
LibWeb: Integrate OffscreenCanvas
